presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationBelowOuterJoin.java

```diff
@@ -0,0 +1,141 @@
+                        aggregationNode,
+                        joinSource.getCriteria(),
+                        joinSource.getFilter(),
+                        joinSource.getLeftHashSymbol(),
+                        joinSource.getRightHashSymbol());
+            }
+
+            if (joinSource.getType() == JoinNode.Type.RIGHT
+                    && !joinSource.getFilter().isPresent()
+                    && groupsOnAllOuterTableColumns(node, joinSource)
+                    && isDistinct(joinSource.getRight())) {
+                ImmutableList.Builder<Symbol> groupingKeysBuilder = ImmutableList.builder();
+                for (JoinNode.EquiJoinClause equiJoinClause : joinSource.getCriteria()) {
+                    groupingKeysBuilder.add(equiJoinClause.getLeft());
+                }
+                AggregationNode aggregationNode = new AggregationNode(
+                        node.getId(),
+                        context.defaultRewrite(joinSource.getLeft()),
+                        node.getAggregations(),
+                        node.getFunctions(),
+                        node.getMasks(),
+                        ImmutableList.of(groupingKeysBuilder.build()),
+                        node.getStep(),
+                        node.getHashSymbol(),
+                        node.getGroupIdSymbol());
+                return new JoinNode(
+                        joinSource.getId(),
+                        joinSource.getType(),
+                        aggregationNode,
+                        context.defaultRewrite(joinSource.getRight()),
+                        joinSource.getCriteria(),
+                        joinSource.getFilter(),
+                        joinSource.getLeftHashSymbol(),
+                        joinSource.getRightHashSymbol());
+            }
+            return context.defaultRewrite(node, context.get());
+        }
+
+        private static boolean groupsOnAllOuterTableColumns(AggregationNode node, JoinNode joinSource)
+        {
+            checkState(joinSource.getType() == JoinNode.Type.LEFT || joinSource.getType() == JoinNode.Type.RIGHT, "expected LEFT or RIGHT JOIN");
+            PlanNode outerNode;
+            if (joinSource.getType().equals(JoinNode.Type.LEFT)) {
+                outerNode = getFirstNonIdentityNode(joinSource.getLeft());
+            }
+            else {
+                outerNode = getFirstNonIdentityNode(joinSource.getRight());
+            }
+            return new HashSet<>(node.getGroupingKeys()).equals(new HashSet<>(outerNode.getOutputSymbols()));
+        }
+
+        private static PlanNode getFirstNonIdentityNode(PlanNode node)
+        {
+            PlanNodeSearcher searcher = new PlanNodeSearcher(node).where(planNode -> !(planNode instanceof ProjectNode) || !((ProjectNode) planNode).isIdentity());
+            return searcher.findFirst().orElseThrow(() -> new IllegalStateException("no nodes below identity project"));
```

Comment:
[minor]

Typically I use `PlanNodeSearcher` in the same way as `Stream`, so I would write this as:

```
return PlanNodeSearcher.searchFrom(node) // probably I would use static import
    .where(planNode -> !(planNode instanceof ProjectNode) || !((ProjectNode) planNode).isIdentity())
    .findFirst()
    .orElseThrow(() -> new IllegalStateException("no nodes below identity project"));
```